### PR TITLE
kernel: thread: make offload_sem visible for releasing it outside

### PR DESCRIPTION
**Pull request by enjiamai**
In order to release irq_offload semaphore outside kernel/thread.c, we make it visible by modifying it non-static under ztest. 
his would be needed such as when call irq_offload() to enter interrupt context and a fatal error happened, then you have to release it in your fatal handler, or the irq_offload will still be locked and no longer be using again.

And what special case will we need this:  for a scenario of some negative test cases.  For example:
1. Mutex cannot be used in interrupt context, so there should be a test case that call k_mutex_lock in interrupt context, 
2. Then it will trigger a assert handler to catch this, and we leverage self-defined post assert handler and call k_thread_abort() or ztest_test_pass() to judge the test case pass.
3. But there is one problem here, the irq_offload semaphore has been taken before irq_offload(), and fatal or assert happened. So before terminated the thread, the resource(offload_sem) shall be released before next test case executing, otherwise, one other test case(or APP) can use offload_load() again, it would be block here.
        
Perhaps another way to achieve this goal  is to add an API interface for release the offload_sem, such as "int release_offload_sem(void)".

I am really not sure what is the better practice for handling this, look forward to your feedback, thanks!   

Signed-off-by: Enjia Mai <enjiax.mai@intel.com>

### CODE REVIEW

**Review by tyrantlizard77 — CHANGES_REQUESTED** [review]
you haven't adequately explained why you need this in the first place

**Comment by enjiamai** [comment]
> you haven't adequately explained why you need this in the first place

Hi~Thanks for your reminder, I add below descriptions into comments. 
And what special case will we need this:  for a scenario of some negative test cases.  For example:
1. Mutex cannot be used in interrupt context, so there should be a test case that call k_mutex_lock in interrupt context, 
2. Then it will trigger a assert handler to catch this, and we leverage self-defined post assert handler and call k_thread_abort() or ztest_test_pass() to judge the test case pass.
3. But there is one problem here, the irq_offload semaphore has been taken before irq_offload(), and fatal or assert happened. So before terminated the thread, the resource(offload_sem) shall be released before next test case executing, otherwise, one other test case(or APP) can use offload_load() again, it would be block here.

**Comment by enjiamai** [comment]
@andrewboie  Or do you know whether Zephyr has some kind of mechanism that can release resource if some a thread hold a resource then fatal error or assert fail happened?
A far as I know, it seems like we have to release them manually. Thanks!

File: kernel/thread.c [hunk 1]
```diff
@@ -908,7 +908,15 @@ static inline int z_vrfy_k_float_disable(struct k_thread *thread)
...
 
 #ifdef CONFIG_IRQ_OFFLOAD
+
+/* Make offload_sem visible outside under testing, in order to release
+ * it outside when error happened.
+ */
+#ifdef CONFIG_ZTEST
+K_SEM_DEFINE(offload_sem, 1, 1);
+#else
 static K_SEM_DEFINE(offload_sem, 1, 1);
```
Comment (andrewboie) [inline kernel/thread.c:918]: you can just delete the static keyword, no need for the ifdef.

I would accept a patch which makes the entire irq_offload() API CONFIG_ZTEST only, to make it clear that this is only for tests.

**Review by tyrantlizard77 — APPROVED** [review]
I see. You want to be able to take a fatal exception inside an irq_offload() handler, but doing so leaves the semaphore locked and nobody else can irq_offload() in the future.

Given that irq_offload() is a testing-only function and not for any kind of application use, I think it is OK to work around in this way.

File: kernel/thread.c (hunk 1 above)
Comment (enjiamai) [reply kernel/thread.c:918]: OK, got it. And I think I would just remove the static. That means not only CONFIG_ZTEST but CONFIG_TEST could use it.  

And may I have you opinion again? I guess there might be few other kinds of resources, would encounter the problem like this during some negative(error) case. If this happened, do you think is it better to have a extra stand-alone CONFIG to get them all visible? such as CONFIG_NEGATIVE_TEST or CONFIG_TEST_ERROR_RECOVER? Thanks!

Comment (andrewboie) [reply kernel/thread.c:918]: I think CONFIG_TEST is enough.

Comment (enjiamai) [reply kernel/thread.c:918]: Got it, thanks!

**Review by LeiW000 — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Pull request merged**